Data/ByteString/Internal/Pure.hs

```diff
@@ -0,0 +1,366 @@
+{-# LANGUAGE MagicHash #-}
+{-# LANGUAGE BangPatterns #-}
+{-# LANGUAGE MultiWayIf #-}
+
+-- | Haskell implementation of C bits
+module Data.ByteString.Internal.Pure
+  ( -- * fpstring.c
+    intersperse
+  , countOcc
+  , countOccBA
+  , reverseBytes
+  , findMaximum
+  , findMinimum
+  , quickSort
+  , elemIndex
+  , isValidUtf8
+  , isValidUtf8BA
+  -- * itoa.c
+  , decodeSignedDec
+  , decodeUnsignedDec
+  , decodeUnsignedDecPadded
+  , decodeUnsignedHex
+  -- * static tables (unaligned!)
+  , lower_hex_table
+  , digit_pairs_table
+  )
+where
+
+import Prelude
+
+import GHC.Exts                 (Ptr(..), ByteArray#, indexWord8Array#, Word8#, Int#, indexWord8OffAddr#)
+import GHC.Types                (Int (..))
+import GHC.Word                 (Word8(..))
+import GHC.Int                  (Int8(..))
+
+import Data.Bits                (Bits(..), shiftR, (.&.))
+import Data.Word
+import Foreign.Ptr              (plusPtr)
+import Foreign.Storable         (Storable(..))
+import Control.Monad            (when)
+
+----------------------------------------------------------------
+-- Haskell version of functions in fpstring.c
+----------------------------------------------------------------
+
+-- | duplicate a string, interspersing the character through the elements of the
+-- duplicated string
+intersperse :: Ptr Word8 -> Ptr Word8 -> Int -> Word8 -> IO ()
+intersperse !dst !src !len !w = case len of
```

Comment:
I don't think `!len` is necessary if the function immediately is scrutinizing `len`

Reply:
I usually do this when I want the arguments to be passed unboxed, whatever I do with them in the body of the function. I think it's clearer for the reader (if I could I would add bangs in the type-signature instead). It's more future-refactoring-proof too because the argument demands won't change even if we change something in the body.